## Statistics exercises

### Exercise 1

In the lady tasting tea experiment, a subject is given eight cups of tea of sample size 8, of which 4 have been prepared with tea added first and four with milk added first. The cups are given in a random order and the subject is asked to determine which cup belongs to which group.

The odds of choosing the correct combination of cups at random are given as 1 in 70 possible combinations or 1.4%.

Task: Calculate the minimum number of cups of tea required to ensure the probability of randomly selecting the correct cups is less than or equal to 1%.

***

The odds of choosing the correct combination of cups increases with sample size *n* and the number of possible combinations, *k* being the number of cups to be chosen (half of *n*). The number of combinations for a sample size *n* can be calculated using the SciPy *comb* package:

In [17]:
from scipy.special import comb

k = 4
n = 8
combinations = comb(n, k, exact=True)
print("Possible combinations of k from n: " + str(combinations))

print("Chance of correct guess: " + str(round(100*(1/combinations),2)) + "%")

Possible combinations of k from n: 70
Chance of correct guess: 1.43%


Increasing *n* will lower the odds of guessing the correct combination at random to less than 1%:

In [18]:
k = 6
n = 12
combinations = comb(n,k, exact=True)
print("Possible combinations of k from n: " + str(combinations))

print("Chance of correct guess: " + str(round(100*(1/combinations),2)) + "%")

Possible combinations of k from n: 924
Chance of correct guess: 0.11%


Selecting 6 cups at random from a sample size of 12 yields 924 possible combinations and a 0.11% chance of guessing correctly.

Lowering *n* to 10 brings the probability closer to the 1% target:

In [19]:
k = 5
n = 10
combinations = comb(n,k, exact=True)
print("Possible combinations of k from n: " + str(combinations))

print("Chance of correct guess: " + str(round(100*(1/combinations),2)) + "%")

Possible combinations of k from n: 252
Chance of correct guess: 0.4%


Because the cups must remain discrete units and *n* and *k* must therefore remain integers, and assuming *k* must be exactly half of *n*, a sample size of 10 brings the probability closest to <=1%.

***

Bonus: How many would be required if you were to let the taster get one cup wrong while maintaining the 1% threshold?

***

Keeping a sample size of *n*=10, *k* is now set to 4 as the subject gets one guess wrong.

In [20]:
k = 4
n = 10
combinations = comb(n,k, exact=True)
print("Possible combinations of k from n: " + str(combinations))

print("Chance of correct guess: " + str(round(100*(1/combinations),2)) + "%")

Possible combinations of k from n: 210
Chance of correct guess: 0.48%


The probability of a correct guess is now at 0.48%, staying under the 1% target. Testing against the other values of *n*:

In [21]:
k = 3
n = 8
combinations = comb(n,k, exact=True)
print("Possible combinations of k from n: " + str(combinations))

print("Chance of correct guess: " + str(round(100*(1/combinations),2)) + "%")

Possible combinations of k from n: 56
Chance of correct guess: 1.79%


In [22]:
k = 5
n = 12
combinations = comb(n,k, exact=True)
print("Possible combinations of k from n: " + str(combinations))

print("Chance of correct guess: " + str(round(100*(1/combinations),2)) + "%")

Possible combinations of k from n: 792
Chance of correct guess: 0.13%


We can see that a sample size of 8 with 3 correct guesses is 1.79%, significantly over the 1% limit, while a sample size of 12 with 5 correct guesses has a probability of 0.13%.

***

### Exercise 2

Task: Use scipy's version of Fisher's exact test to simulate the Lady Tasting Tea problem.

***

Fisher's exact test is a statistical significance test used in the analysis of contingency tables, developed as part of the lady tasting tea experiment.

Sticking with the sample size of 10 from above, table describing a correct guess of five cups with tea added first and five with milk added first would be:

In [49]:
table = [[5,0],
         [0,5]]

Passing this table to Fisher's exact test, we return a p-value of 0.8%, which is double the expected value for the sample size:

In [58]:
from scipy.stats import fisher_exact

fisher_exact(table)

(inf, 0.007936507936507929)

This is because *fisher_exact()* contains the optional parameter "alternative", which defines the conditions for the alternative (non-null) hypothesis. The default value of the parameter is "two-sided", counting both tails of the probability distribution, i.e., the probability that the odds ratio of the population underlying the observation is not one, irrespective of the value being greater or lesser than one.

To get the p-value of the table describing a correct guess (0.4%, as found in exercise 1), "alternative" must be set to "greater":

In [59]:
fisher_exact(table, alternative="greater")

(inf, 0.0039682539682539646)

The p-value now conforms to the prediction found in exercise 1, that the ([4,0],[0,4]) table describing a correct guess has a probability of 0.4%.

Because p < 0.05, we can reject the null hypothesis in this instance.

***

### Exercise 3

Task: Take the code from the Examples section of the scipy stats documentation for independent samples t-tests, add it to your own notebook and explain how it works using MarkDown cells and code comments. Improve it in any way you think it could be improved.

***

A t-test is any statistical hypothesis test to compare the means of two groups of samples. The sample code explores different possible conditions in which the function may be used, such as varying scales, sample sizes, etc.


The sample code begins by importing the required packages and setting up the number generator taken from numpy:

In [65]:
# numpy for creating the random number generator from the np.random package
import numpy as np

# for the ttest functions
from scipy import stats

# random number generator
rng = np.random.default_rng()

It then runs a test using two normally-distributed samples (this is one of the assumptions of the t-test) with identical sample sizes of values taken from the normal distribution, means (loc), and standard deviation (scale), with a random factor added:

In [84]:
# stats.norm.rvs will calculate the normal distribution for each sample
# ttest_ind() will then perform the t-test on those values

rvs1 = stats.norm.rvs(loc=5, scale=10, size=500, random_state=rng)
rvs2 = stats.norm.rvs(loc=5, scale=10, size=500, random_state=rng)

stats.ttest_ind(rvs1, rvs2)

Ttest_indResult(statistic=0.9872987363918337, pvalue=0.3237355074159033)

The *ttest_ind()* function returns the calculated t-statistic of 0.99, and a p-value of 32%.

These results indicate that the null hypothesis should not be rejected due to the p-value being greater than 0.05.

In the next line, the *equal_var* parameter is set to *False*. This parameter determines whether the *ttest_ind()* function should assume equal variances between rvs1 and rvs2, i.e., that the degree to which a random variable from one sample will differ from that sample's expected value according to the distribution, will be equal to that from the other sample.

The parameter is set to *True* by default, in which case the function assumes equal variances between the samples and performs a standard independent 2-sample test. If set to *False*, it assumes unequal variances and performs Welch's t-test, which is more reliable with unequal variances and  potentially unequal sample sizes. Welch's t-test still assumes normality in the samples.

In [85]:
stats.ttest_ind(rvs1, rvs2, equal_var=False)

Ttest_indResult(statistic=0.9872987363918337, pvalue=0.3237356915530597)

In this case the t-statistic and p-value are nearly identical, showing that the variances between the samples are the same and Welch's test is not a good fit for these samples.

The code now creates a new sample, rvs3, with loc=5, scale=20, and size=500, plus the random factor. It then performs a t-test on rvs1 and rvs3, with equal variances assumed by *equal_vr* being left to its default value.

In [87]:
rvs3 = stats.norm.rvs(loc=5, scale=20, size=500, random_state=rng)
stats.ttest_ind(rvs1, rvs3)

Ttest_indResult(statistic=0.5506682852081517, pvalue=0.5819843273483702)

Comparing the means of the two samples, *ttest_ind()* finds a t-statistic of 0.55 and a p-value of 58%.

In [88]:
stats.ttest_ind(rvs1, rvs3, equal_var=False)

Ttest_indResult(statistic=0.5506682852081517, pvalue=0.5820290227405511)

With the unequal variances, the results are again marginally different.

In the step of the sample code, a fourth sample is declared with a significantly lower sample size:

In [91]:
rvs4 = stats.norm.rvs(loc=5, scale=20, size=100, random_state=rng)
stats.ttest_ind(rvs1, rvs4)

Ttest_indResult(statistic=2.330566916705046, pvalue=0.02010783713446677)

In [92]:
stats.ttest_ind(rvs1, rvs4, equal_var=False)

Ttest_indResult(statistic=1.4634849320028398, pvalue=0.14625158080122577)

The difference in sample size has led to a significant difference in the t-statistic and p-value when assuming equal or unequal variances, although neither has a low enough p-value for the null hypothesis to be rejected.

Another sample is declared with different means, variance and sample size:

In [93]:
rvs5 = stats.norm.rvs(loc=8, scale=20, size=100, random_state=rng)
stats.ttest_ind(rvs1, rvs5)

Ttest_indResult(statistic=-1.6550938908081205, pvalue=0.09843013463975184)

In [94]:
stats.ttest_ind(rvs1, rvs5, equal_var=False)

Ttest_indResult(statistic=-1.0712362010251228, pvalue=0.2864363162731875)

When performing a permutation test, more permutations typically yields more accurate results. The *random_state* attribute ("None" by default, which uses the np.random.RandomState singleton) is set to the generator decalred at the beginning of the code:

In [95]:
stats.ttest_ind(rvs1, rvs5, permutations=10000,
                random_state=rng)

Ttest_indResult(statistic=-1.6550938908081205, pvalue=0.098)

The code next declares two new samples, one of which has a longer tail than the other, and uses the *Trim* keyword to set a value of 20% trimming


In [96]:
a = (56, 128.6, 12, 123.8, 64.34, 78, 763.3)
b = (1.1, 2.9, 4.2)

Use the trim keyword to perform a trimmed (Yuen) t-test. For example, using 20% trimming, trim=.2, the test will reduce the impact of one element from each tail of sample a. It will have no effect on sample b because np.floor(trim*len(b)) is 0.

In [97]:
stats.ttest_ind(a, b, trim=.2)

Ttest_indResult(statistic=3.4463884028073513, pvalue=0.01369338726499547)